### ISSUE TITLE
Accessibility for Visually Impaired Users

### ISSUE THREAD

**Issue by DibyaXPP** [aui-framework/aui#347]
Applications built with AUI cannot be read by NVDA or other screen readers on Windows, making AUI applications inaccessible for people who cannot see or have vision problems.

UI elements such as buttons, labels, lists, and checkboxes can be integrated with COM automation, utilizing the Microsoft Active Accessibility components found primarily in the following header: [OLEACC](https://learn.microsoft.com/en-us/windows/win32/api/oleacc/). This can make AUI accessible to people who are visually impaired.

Please consider this feature request. There are not many cross-platform GUI libraries that provide support for people with vision challenges, many of whom are software developers and engineers like us.

**Comment by Alex2772** [comment]
Thank you for your detailed description.

We'd take this into consideration.

> There are not many cross-platform GUI 

As far as I know, Qt supports OLEACC.

**Comment by Alex2772** [comment]
https://github.com/qt/qt/blob/4.8/src/gui/kernel/qapplication_win.cpp `WM_GETOBJECT`

**Comment by DibyaXPP** [comment]
Thanks for looking into it

**Comment by DibyaXPP** [comment]
Hi @Alex2772  I found something which might solve accessibility issue across the platform [though i am mainly interested in Windows].
https://github.com/AccessKit/accesskit 
It has C Binding, Since it is licensed under MIT, i will personally prefer it being ultimately rewritten to Modern C++ instead of RUST, that will allow seamless build and portability to other C++ Library. 
I won't mind volunteering for rewrite if u look into feasibility of integrating it. 
May be we can call it AccessKit++

**Comment by Alex2772** [comment]
Hi @DibyaXPP,

Thanks, it is a good example.

I'd rather prefer to avoid dependencies and implement accessibility as AUI's feature.

> feasibility of integrating it.

There's no showstoppers from AUI in regards to accessibility API. The only thing that may seem complicated is IME (text field typing) which is currently somewhat unstable in AUI. I'm planning to put accessibility implementation somewhere after AUI's IME stabilizes. AUI has D-Bus bindings so it can be used on Linux systems.

**Comment by Alex2772** [comment]
> I won't mind volunteering

I'm planning doing it by myself, but I may ask you some questions during designing as I'm looking at accessibility API from perspective of UI automation for testing in the first place.

You can focus on writing documentation for beginners as this is the exact thing I can't do.

**Comment by DibyaXPP** [comment]
I'm planning doing it by myself, but I may ask you some questions during designing as I'm looking at accessibility API from perspective of UI automation for testing in the first place.
=> A bit of request, Whenever u test out things, specially keep a detailed documentation for it since that will useful to make 
other application using GLEW/SFML accessible to users.

**Comment by Alex2772** [comment]
> make other application

I'm not sure I can support non-AUI-based applications, since I'll quickly depend on AUI probably.

**Comment by DataTriny** [comment]
Hello, AccessKit developer here.

@Alex2772 If we had C++ bindings, would that make you reconsider implementing accessibility yourself? I'm asking because this is something we want to have eventually.

If we can do something to improve the build process for your usecase, please let me know.

**Comment by Alex2772** [comment]
> Hello, AccessKit developer here.
> 
> @Alex2772 If we had C++ bindings, would that make you reconsider implementing accessibility yourself? I'm asking because this is something we want to have eventually.
> 
> If we can do something to improve the build process for your usecase, please let me know.

@DataTriny AccessKit has C bindings so it can be used in C++ projects already. The point is AccessKit is written in Rust. Of course, there's corrosion, so we can use AccessKit in С/C++ project.  But we ultimately adding another huge dependency: one needs a C++ compiler and **a Rust compiler** to build AUI. Obvious question: "AUI is a C++ library, why does it need a Rust compiler??" - "Yeah AUI is a C++ library, but we use Rust library to implement accessibility features. Sorry!"

I'm, however, curious to test corrosion.

**Comment by DataTriny** [comment]
I just wanted to make sure that dealing with non idiomatic/lower-level API from C++ was not the issue here.

Good luck with your implementation.

**Comment by DataTriny** [comment]
The Godot game engine is set to [use AccessKit](https://github.com/godotengine/godot/pull/76829) as well.

It is important to ask yourself whether you really have the resources to maintain a cross-platform accessibility provider. You might come to the conclusion that adding a Rust dependency is a simpler option.

If you still want to do it on your own, I don't recommend going with Microsoft Active Accessibility as mentioned in this thread previously. It has been abandoned by Microsoft a while ago.

Happy to answer any question you might have regarding an AccessKit integration.
